### DISCUSSION TITLE
Unreadable PDF content

### DISCUSSION THREAD

**Discussion by CagliotiAtik** [spatie/laravel-pdf#173] [Q&A]
# What happened?

Spatie LaravelPDF works well on windows machines, but in the Debian12 production all the PDF generated are unreadable.

First of all, I had a previous problem that completely prevented me from generating PDFs. I solved it by adopting the solution proposed in this [discussion](https://github.com/spatie/laravel-pdf/discussions/55#discussioncomment-9586073).

Now the related problem is that I cannot generate a valid PDF.
Every time I try to open one I get a size other than 0 but (on both Edge and Firefox) I get a “Unable to open document” error.

## How to reproduce the bug

I am going to generate the document from a Laravel blade file.
I have created a Service for generating and saving PDFs to which I pass two variables:

- $view, the name of the view to “transform”
- $values, an array of variables to pass to the view

The code that goes to perform the generation and saving is divided into two files:

```PHP
// FILE 1 => PDFGenerationService

// for the PDF generation
public function generatePDF(string $view, array $values): void
    {
        $this->pdf = Pdf::view(
            $view,
            $values
        )
            ->orientation($this->orientation)
            ->format($this->format)
            ->margins(
                top: $this->topMargin,
                right: $this->rightMargin,
                bottom: $this->bottomMargin,
                left: $this->leftMargin
            );
    }

//for the PDF disk save
public function savePDF(string $disk, string $path): PdfBuilder
    {
        return $this->pdf
            ->disk($disk)
            ->save($path);
    }


//FILE 2 => GenerationController
private function generatePDF(string $uniqueName, Practice $practice): array
    {

        $acceptanceFormInfo = $practice->acceptanceForm()->first();
        $userInfo = $practice->user()->first();

        $values = [
            .....
        ];

        $this->PDFGenerationService->generatePDF(
            $this->templateTypes[2]['view'],
            $values
        );
        $this->PDFGenerationService->savePDF(
            'practice',
            $uniqueName
        );

        return [
            'filePath' => Storage::disk('practice')->path($uniqueName),
            'fileUrl' => Storage::disk('practice')->url($uniqueName),
        ];

    }
```
The file generated after downloading it (but it is the same size if I don't download it)
![immagine](https://github.com/user-attachments/assets/34f516d1-6755-49e3-9d65-7105666ff1a0)
![immagine](https://github.com/user-attachments/assets/8c9392e9-23d6-4926-8757-1200a73f7afb)

And the result if i try to open it in Edge
![immagine](https://github.com/user-attachments/assets/3faf7e9a-5461-48c6-b7ab-af0e35e75e4b)

Anyone who can give me guidance on how I can solve this problem?

# Verisons
## Package Version
![immagine](https://github.com/user-attachments/assets/beafd5ad-f7bd-484c-aff5-f8fc763ddb2f)

## PHP Version
![immagine](https://github.com/user-attachments/assets/37a95ae4-e219-4eb1-8ebe-182021c8bdf3)

## Laravel Version
11.9

## Operating System
Kernel: 6.1.0-12-amd64

PRETTY_NAME="Debian GNU/Linux 12 (bookworm)"
NAME="Debian GNU/Linux"
VERSION_ID="12"
VERSION="12 (bookworm)"
VERSION_CODENAME=bookworm
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"

# Notes
As mentioned, the solution works perfectly with development Windows machines.
In fact, for them I did not need to modify the puppeteer cache, indeed, if I try to insert the file with the modified configuration on Windows machines I get the following error:

```
The command "node "C:\Projects\XXXX\vendor\spatie\browsershot\src/../bin/browser.cjs" "{""url"":""file:\/\/C:\\Users\\FD635~1.CAG\\AppData\\Local\\Temp\\1179708237-0030941001724079467\\index.html"",""action"":""pdf"",""options"":{""args"":[],""viewport"":{""width"":800,""height"":600},""displayHeaderFooter"":false,""margin"":{""top"":""10mm"",""right"":""10mm"",""bottom"":""10mm"",""left"":""10mm""},""format"":""a4"",""printBackground"":true}}"" failed.

Exit Code: 1(General error)

Working directory: C:\Projects\XXXX\public

Output:
================


Error Output:
================
Error: Could not find Chrome (ver. 127.0.6533.72). This can occur if either
 1. you did not perform an installation before running the script (e.g. `npx puppeteer browsers install chrome`) or
 2. your cache path is incorrectly configured (which is: C:\Projects\XXXX\.cache\puppeteer).
For (2), check out our guide on configuring puppeteer at https://pptr.dev/guides/configuration.
    at ChromeLauncher.resolveExecutablePath (C:\Projects\XXXX\node_modules\puppeteer-core\lib\cjs\puppeteer\node\ProductLauncher.js:291:27)
    at ChromeLauncher.executablePath (C:\Projects\XXXX\node_modules\puppeteer-core\lib\cjs\puppeteer\node\ChromeLauncher.js:209:25)
    at ChromeLauncher.computeLaunchArguments (C:\Projects\XXXX\node_modules\puppeteer-core\lib\cjs\puppeteer\node\ChromeLauncher.js:89:37)
    at async ChromeLauncher.launch (C:\Projects\XXXX\node_modules\puppeteer-core\lib\cjs\puppeteer\node\ProductLauncher.js:70:28)
    at async callChrome (C:\Projects\XXXX\vendor\spatie\browsershot\bin\browser.cjs:92:23)

```

**Comment by bluesheep100** [comment]
There is currently a known issue in Browsershot (dependency of laravel-pdf) that prevents it from working with Puppeteer v23+.
The problem arises because Puppeteer modified how data is returned from the browser instance, which breaks browsershot.
Until it's fixed, ensure you're on spatie/browsershot 4.2.1, and Puppeteer 22.15.0; that should get it working.
